File: linux/src/meeting_recorder/ui/engine_proxy.py
"""
Client-side proxy to the daemon's Engine (the window process talks through this).

Wraps the ``io.github.jmarceno.Gravaai.Engine`` D-Bus interface: the
window issues commands as method calls and receives state as ``SnapshotChanged``
(and Error/Output/OpenUseExisting/PresentWindow) signals. Command calls are
fire-and-forget (async); the few that need a value (snapshot, job folder,
summarize) are synchronous.

GTK-free itself (``Gio``/``GLib`` only); it is constructed by ``window_app.py``
and handed to ``MainWindow``. Not unit-tested — it needs a live bus and the
running daemon on the other end.
"""

from __future__ import annotations

import logging
from collections.abc import Callable

from gi.repository import Gio, GLib

from ..core.daemon_watch import should_exit_on_owner_change
from ..daemon.dbus_service import ENGINE_IFACE, ENGINE_NAME, ENGINE_PATH

logger = logging.getLogger(__name__)


class EngineProxy:
    def __init__(
        self,
        *,
        on_snapshot: Callable[[str], None],
        on_error: Callable[[str], None],
        on_output: Callable[[str], None],
        on_open_use_existing: Callable[[], None],
        on_present: Callable[[], None],
        on_daemon_gone: Callable[[], None] = lambda: None,
    ) -> None:
        self._on_snapshot = on_snapshot
        self._on_error = on_error
        self._on_output = on_output
        self._on_open_use_existing = on_open_use_existing
        self._on_present = on_present
        self._on_daemon_gone = on_daemon_gone
        # Install signals go to whatever Settings/Models page is currently open;
        # it (un)registers itself here since it is created and destroyed on demand.
        self._install_listeners: list = []
        self._conn = Gio.bus_get_sync(Gio.BusType.SESSION, None)
        self._conn.signal_subscribe(
            ENGINE_NAME,
            ENGINE_IFACE,
            None,  # all signals
            ENGINE_PATH,
            None,
            Gio.DBusSignalFlags.NONE,
            self._on_signal,
        )
        # A window must never outlive its daemon: watch the Engine bus name and
        # exit if the daemon that spawned us quits or crashes, so a hidden
        # (kept-in-memory) window can't linger and double up on the next daemon's
        # PresentWindow broadcast. See core/daemon_watch.py.
        self._daemon_seen = False
        self._watch_id = Gio.bus_watch_name_on_connection(
            self._conn,
            ENGINE_NAME,
            Gio.BusNameWatcherFlags.NONE,
            self._on_daemon_appeared,
            self._on_daemon_vanished,
        )

    # ------------------------------------------------------------------
    # Daemon liveness (bus-name owner) watch
    # ------------------------------------------------------------------

    def _on_daemon_appeared(self, _conn, _name, _owner) -> None:
        self._daemon_seen = True

    def _on_daemon_vanished(self, _conn, _name) -> None:
        if should_exit_on_owner_change(self._daemon_seen, has_owner=False):
            logger.info("Daemon gone — window exiting to avoid orphaning")
            self._on_daemon_gone()

    # ------------------------------------------------------------------
    # Signals from the daemon
    # ------------------------------------------------------------------

    def _on_signal(self, _conn, _sender, _path, _iface, signal_name, params):
        if signal_name == "SnapshotChanged":
            self._on_snapshot(params.unpack()[0])
        elif signal_name == "Error":
            self._on_error(params.unpack()[0])
        elif signal_name == "Output":
            self._on_output(params.unpack()[0])
        elif signal_name == "OpenUseExisting":
            self._on_open_use_existing()
        elif signal_name == "PresentWindow":
            self._on_present()
        elif signal_name == "InstallProgress":
            key, text = params.unpack()
            for listener in list(self._install_listeners):
                listener.on_install_progress(key, text)
        elif signal_name == "InstallFinished":
            key, ok, message = params.unpack()
            for listener in list(self._install_listeners):
                listener.on_install_finished(key, ok, message)

    # ------------------------------------------------------------------
    # Command calls
    # ------------------------------------------------------------------

    def _call(self, method: str, body=None) -> None:
        self._conn.call(
            ENGINE_NAME,
            ENGINE_PATH,
            ENGINE_IFACE,
            method,
            body,
            None,
            Gio.DBusCallFlags.NONE,
            -1,
            None,
            None,  # fire-and-forget
        )

    def _call_sync(self, method: str, body=None, reply_type: str = "(s)"):
        try:
            result = self._conn.call_sync(
                ENGINE_NAME,
                ENGINE_PATH,
                ENGINE_IFACE,
                method,
                body,
                GLib.VariantType.new(reply_type),
                Gio.DBusCallFlags.NONE,
                -1,
                None,
            )
            return result.unpack()
        except GLib.Error as exc:
            logger.error("Engine.%s failed: %s", method, exc)
            return None

    # --- recording lifecycle ---
    def start_recording(self, mode: str) -> None:
        self._call("StartRecording", GLib.Variant("(s)", (mode,)))

    def set_title(self, title: str) -> None:
        self._call("SetTitle", GLib.Variant("(s)", (title,)))

    def pause(self) -> None:
        self._call("Pause")

    def resume(self) -> None:
        self._call("Resume")

    def stop(self) -> None:
        self._call("Stop")

    def cancel_countdown(self) -> None:
        self._call("CancelCountdown")

    def cancel_save(self) -> None:
        self._call("CancelSave")

    def cancel(self) -> None:
        self._call("Cancel")

    # --- jobs ---
    def import_existing(self, audio: str, transcript: str, notes: str, label: str) -> None:
        self._call("ImportExisting", GLib.Variant("(ssss)", (audio, transcript, notes, label)))

    def summarize_meeting(self, audio: str, transcript: str, notes: str, label: str) -> str:
        res = self._call_sync(
            "SummarizeMeeting", GLib.Variant("(ssss)", (audio, transcript, notes, label))
        )
        return res[0] if res else ""

    def cancel_job(self, job_id: int) -> None:
        self._call("CancelJob", GLib.Variant("(i)", (job_id,)))

    def retry_job(self, job_id: int) -> None:
        self._call("RetryJob", GLib.Variant("(i)", (job_id,)))

    def dismiss_job(self, job_id: int) -> None:
        self._call("DismissJob", GLib.Variant("(i)", (job_id,)))

    def job_folder(self, job_id: int) -> str:
        res = self._call_sync("JobFolder", GLib.Variant("(i)", (job_id,)))
        return res[0] if res else ""

    def output_folder(self) -> str:
        res = self._call_sync("OutputFolder")
        return res[0] if res else ""

    # --- model/engine installs (run in the daemon) ---
    def add_install_listener(self, listener) -> None:
        self._install_listeners.append(listener)

    def remove_install_listener(self, listener) -> None:
        if listener in self._install_listeners:
            self._install_listeners.remove(listener)

    def start_install(self, spec_json: str) -> None:
        self._call("StartInstall", GLib.Variant("(s)", (spec_json,)))

    def get_installs(self) -> str:
        res = self._call_sync("GetInstalls")
        return res[0] if res else "[]"

    # --- misc ---
    def reload_config(self) -> None:
        self._call("ReloadConfig")

    def open_window(self) -> None:
        self._call("OpenWindow")

    def get_snapshot(self) -> str:
        res = self._call_sync("GetSnapshot")
        return res[0] if res else ""

    def quit(self) -> None:
        self._call("Quit")

File: linux/src/ui/engine_proxy.rs
//! D-Bus client for the Engine service.
//!
//! The window process renders snapshots fetched over D-Bus (`GetSnapshot`,
//! kept fresh by `SnapshotChanged`) and forwards clicks back as method calls.
//! Fire-and-forget for commands; getters await replies.

use crate::core::daemon_watch::should_exit_on_owner_change;
use crate::daemon::dbus_service::ENGINE_NAME;

/// Generated proxy struct is `EngineProxy` (from trait `Engine`).
#[zbus::proxy(
    interface = "io.github.jmarceno.GravaAi.Engine",
    default_service = "io.github.jmarceno.GravaAi",
    default_path = "/io/github/jmarceno/GravaAi"
)]
pub trait Engine {
    fn start_recording(&self, mode: &str) -> zbus::Result<()>;
    fn start_custom_recording(&self, devices_json: &str) -> zbus::Result<String>;
    fn set_title(&self, title: &str) -> zbus::Result<()>;
    fn pause(&self) -> zbus::Result<()>;
    fn resume(&self) -> zbus::Result<()>;
    fn stop(&self) -> zbus::Result<()>;
    fn cancel_countdown(&self) -> zbus::Result<()>;
    #[zbus(name = "CancelSave")]
    fn cancel_save(&self) -> zbus::Result<()>;
    #[zbus(name = "Cancel")]
    fn cancel(&self) -> zbus::Result<()>;
    fn import_existing(
        &self,
        audio: &str,
        transcript: &str,
        notes: &str,
        label: &str,
    ) -> zbus::Result<()>;
    fn summarize_meeting(
        &self,
        audio: &str,
        transcript: &str,
        notes: &str,
        label: &str,
    ) -> zbus::Result<String>;
    fn transcribe_meeting(
        &self,
        audio: &str,
        transcript: &str,
        notes: &str,
        label: &str,
    ) -> zbus::Result<String>;
    fn cancel_job(&self, id: i32) -> zbus::Result<()>;
    fn retry_job(&self, id: i32) -> zbus::Result<()>;
    fn dismiss_job(&self, id: i32) -> zbus::Result<()>;
    fn job_folder(&self, id: i32) -> zbus::Result<String>;
    fn output_folder(&self) -> zbus::Result<String>;
    fn reload_config(&self) -> zbus::Result<()>;
    fn open_window(&self) -> zbus::Result<()>;
    fn get_snapshot(&self) -> zbus::Result<String>;
    fn start_install(&self, spec: &str) -> zbus::Result<()>;
    fn get_installs(&self) -> zbus::Result<String>;
    fn quit(&self) -> zbus::Result<()>;

    #[zbus(signal)]
    fn snapshot_changed(&self, json: String) -> zbus::Result<()>;

    #[zbus(signal)]
    fn error(&self, msg: String) -> zbus::Result<()>;

    #[zbus(signal)]
    fn output(&self, text: String) -> zbus::Result<()>;

    #[zbus(signal)]
    fn open_use_existing(&self) -> zbus::Result<()>;

    #[zbus(signal)]
    fn present_window(&self) -> zbus::Result<()>;

    #[zbus(signal)]
    fn install_progress(&self, key: String, text: String) -> zbus::Result<()>;

    #[zbus(signal)]
    fn install_finished(&self, key: String, ok: bool, message: String) -> zbus::Result<()>;
}

// The macro above generates the `EngineProxy` D-Bus client struct.

// ---------------------------------------------------------------------------
// UI-friendly handle: fire-and-forget commands + async getters, shared
// across UI callbacks. The zbus connection is leaked ('static) since it lives
// as long as the window process.
// ---------------------------------------------------------------------------

/// Install progress/finished events forwarded to the open Models page.
#[derive(Debug, Clone)]
pub enum InstallUiEvent {
    Progress(String, String),
    Finished(String, bool, String),
}

#[derive(Clone)]
pub struct ProxyHandle {
    proxy: EngineProxy<'static>,
    rt: tokio::runtime::Handle,
}

impl ProxyHandle {
    pub fn new(conn: &'static zbus::Connection, rt: tokio::runtime::Handle) -> zbus::Result<Self> {
        let rt_c = rt.clone();
        Ok(Self {
            proxy: rt.block_on(async move { EngineProxy::new(conn).await })?,
            rt: rt_c,
        })
    }

    fn spawn<F>(&self, fut: F)
    where
        F: std::future::Future<Output = ()> + Send + 'static,
    {
        self.rt.spawn(fut);
    }

    // --- recording lifecycle (fire-and-forget) ---
    pub fn start_recording(&self, mode: &str) {
        let (p, mode) = (self.proxy.clone(), mode.to_string());
        self.spawn(async move {
            if let Err(e) = p.start_recording(&mode).await {
                log::error!("Engine.StartRecording failed: {e:#}");
            }
        });
    }

    pub fn set_title(&self, title: &str) {
        let (p, title) = (self.proxy.clone(), title.to_string());
        self.spawn(async move {
            if let Err(e) = p.set_title(&title).await {
                log::error!("Engine.SetTitle failed: {e:#}");
            }
        });
    }

    pub fn pause(&self) {
        let p = self.proxy.clone();
        self.spawn(async move {
            if let Err(e) = p.pause().await {
                log::error!("Engine.Pause failed: {e:#}");
            }
        });
    }

    pub fn resume(&self) {
        let p = self.proxy.clone();
        self.spawn(async move {
            if let Err(e) = p.resume().await {
                log::error!("Engine.Resume failed: {e:#}");
            }
        });
    }

    pub fn stop(&self) {
        let p = self.proxy.clone();
        self.spawn(async move {
            if let Err(e) = p.stop().await {
                log::error!("Engine.Stop failed: {e:#}");
            }
        });
    }

    pub fn cancel_countdown(&self) {
        let p = self.proxy.clone();
        self.spawn(async move {
            if let Err(e) = p.cancel_countdown().await {
                log::error!("Engine.CancelCountdown failed: {e:#}");
            }
        });
    }

    pub fn cancel_save(&self) {
        let p = self.proxy.clone();
        self.spawn(async move {
            if let Err(e) = p.cancel_save().await {
                log::error!("Engine.CancelSave failed: {e:#}");
            }
        });
    }

    pub fn cancel(&self) {
        let p = self.proxy.clone();
        self.spawn(async move {
            if let Err(e) = p.cancel().await {
                log::error!("Engine.Cancel failed: {e:#}");
            }
        });
    }

    // --- jobs ---
    pub fn import_existing(&self, audio: &str, transcript: &str, notes: &str, label: &str) {
        let (p, a, t, n, l) = (
            self.proxy.clone(),
            audio.to_string(),
            transcript.to_string(),
            notes.to_string(),
            label.to_string(),
        );
        self.spawn(async move {
            if let Err(e) = p.import_existing(&a, &t, &n, &l).await {
                log::error!("Engine.ImportExisting failed: {e:#}");
            }
        });
    }

    pub fn cancel_job(&self, job_id: i64) {
        let p = self.proxy.clone();
        self.spawn(async move {
            if let Err(e) = p.cancel_job(job_id as i32).await {
                log::error!("Engine.CancelJob failed: {e:#}");
            }
        });
    }

    pub fn retry_job(&self, job_id: i64) {
        let p = self.proxy.clone();
        self.spawn(async move {
            if let Err(e) = p.retry_job(job_id as i32).await {
                log::error!("Engine.RetryJob failed: {e:#}");
            }
        });
    }

    pub fn dismiss_job(&self, job_id: i64) {
        let p = self.proxy.clone();
        self.spawn(async move {
            if let Err(e) = p.dismiss_job(job_id as i32).await {
                log::error!("Engine.DismissJob failed: {e:#}");
            }
        });
    }

    // --- installs ---
    pub fn start_install(&self, spec_json: &str) {
        let (p, spec) = (self.proxy.clone(), spec_json.to_string());
        self.spawn(async move {
            if let Err(e) = p.start_install(&spec).await {
                log::error!("Engine.StartInstall failed: {e:#}");
            }
        });
    }

    pub fn reload_config(&self) {
        let p = self.proxy.clone();
        self.spawn(async move {
            if let Err(e) = p.reload_config().await {
                log::error!("Engine.ReloadConfig failed: {e:#}");
            }
        });
    }

    // --- async getters (called by the worker, never the Qt thread) ---
    fn block_on<F, T>(&self, fut: F) -> Option<T>
    where
        F: std::future::Future<Output = zbus::Result<T>>,
    {
        match self.rt.block_on(fut) {
            Ok(v) => Some(v),
            Err(e) => {
                log::error!("Engine call failed: {e:#}");
                None
            }
        }
    }

    pub fn get_snapshot(&self) -> String {
        let p = self.proxy.clone();
        self.block_on(async move { p.get_snapshot().await })
            .unwrap_or_default()
    }

    pub fn output_folder(&self) -> String {
        let p = self.proxy.clone();
        self.block_on(async move { p.output_folder().await })
            .unwrap_or_default()
    }

    pub fn summarize_meeting(
        &self,
        audio: &str,
        transcript: &str,
        notes: &str,
        label: &str,
    ) -> String {
        let (p, a, t, n, l) = (
            self.proxy.clone(),
            audio.to_string(),
            transcript.to_string(),
            notes.to_string(),
            label.to_string(),
        );
        self.block_on(async move { p.summarize_meeting(&a, &t, &n, &l).await })
            .unwrap_or_default()
    }

    pub fn get_installs(&self) -> String {
        let p = self.proxy.clone();
        self.block_on(async move { p.get_installs().await })
            .unwrap_or_else(|| "[]".to_string())
    }

    pub fn proxy(&self) -> &EngineProxy<'static> {
        &self.proxy
    }
}

/// Watch the Engine bus name; invoke `on_gone` when the daemon vanishes —
/// but only after the name was seen owned, so a startup race can't kill the
/// window early (see [`should_exit_on_owner_change`]).
pub async fn watch_daemon_owner(conn: zbus::Connection, on_gone: impl FnOnce() + Send + 'static) {
    let mut seen_owned = false;
    let proxy = match zbus::fdo::DBusProxy::new(&conn).await {
        Ok(p) => p,
        Err(_) => return,
    };
    // Poll (simple + robust across zbus versions; 2s cadence is plenty for a
    // daemon-lifetime watch).
    loop {
        tokio::time::sleep(std::time::Duration::from_secs(2)).await;
        let Ok(bus_name) = zbus::names::BusName::try_from(ENGINE_NAME) else {
            return;
        };
        let has_owner: bool = proxy.name_has_owner(bus_name).await.unwrap_or(true);
        if has_owner {
            seen_owned = true;
        } else if should_exit_on_owner_change(seen_owned, has_owner) {
            on_gone();
            return;
        }
    }
}
